interfaces/builtin/usb_gadget.go

```diff
@@ -29,25 +48,214 @@ const usbGadgetBaseDeclarationSlots = `
     deny-auto-connection: true
 `
 
-const usbGadgetConnectedPlugAppArmor = `
-# https://www.kernel.org/doc/Documentation/usb/gadget_configfs.txt
-# Allow creating new gadgets under usb_gadget, which is creating
-# new directories
-/sys/kernel/config/usb_gadget/ rw,
-# Allow creating sub-directories, symlinks and files under those
-# directories
-/sys/kernel/config/usb_gadget/** rw,
-
-# Allow access to UDC
-/sys/class/udc/ r,
+const usbGadgetConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access. No filtering here, as we
+# rely on AppArmor to filter the mount operations.
+mount
+umount
+umount2
 `
 
+// usbGadgetInterface allows creating transient and persistent mounts
+type usbGadgetInterface struct {
+	commonInterface
+}
+
+type ffsMountInfo struct {
+	name       string
+	where      string
+	persistent bool
+}
+
+// Until we have a clear picture of how this should work, disallow creating
+// persistent mounts into $SNAP_DATA or $SNAP_USER_DATA
+func validatePersistentWhere(where string) error {
+	if strings.HasPrefix(where, "$SNAP_DATA") ||
+		strings.HasPrefix(where, "$SNAP_USER_DATA") {
+		return errors.New(`usb-gadget "persistent" attribute cannot be used to mount onto $SNAP_DATA or $SNAP_USER_DATA`)
+	}
+	return nil
+}
+
+func (mi *ffsMountInfo) validate() error {
+	// for ffs the name is the name of the function, which is not a path so we
+	// just need to ensure it doesn't contain any AppArmor regex characters.
+	if err := validateNoAppArmorRegexpWithError(`cannot use usb-gadget "name" attribute`, mi.name); err != nil {
+		return err
+	}
+
+	// Reuse the where validation from mount-control, as the semantics are the same
+	if err := validateWhereAttr(mi.where); err != nil {
+		return err
+	}
+
+	if mi.persistent {
+		if err := validatePersistentWhere(mi.where); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ffsMounts(plug interfaces.Attrer) ([]map[string]any, error) {
+	var mounts []map[string]any
+	err := plug.Attr("ffs-mounts", &mounts)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return nil, mountAttrTypeError
+	}
+	return mounts, nil
+}
+
+func enumerateFFSMounts(mounts []map[string]any, yield func(*ffsMountInfo, error) bool) {
+	for _, mount := range mounts {
+		name, ok := mount["name"].(string)
+		if !ok {
+			yield(nil, fmt.Errorf(`usb-gadget FunctionFS mount "name" must be a string`))
```

Comment:
there does not seem to be a use case for doing the fancy thing,so wouldn't this be simpler?
```
func enumerateFFSMounts(mounts []map[string]any, yield func(*ffsMountInfo) error) error {
// ...
	where, ok := mount["where"].(string)
	if !ok {
		return fmt.Errorf(`usb-gadget FunctionFS mount "where" must be a string`)
	}
...
	if err := yield(mountInfo, nil); err != nil {
			return err
	} 
}
```
and it would be consistent with `enumerateMounts()`

Reply:
Yea this ended up being much simpler, thank you for this input :)